endless list

masya38ru opened the pull request:
походу я в ударе

masya38ru committed:
endless list

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,70 @@
+import retrofit2.Callback
+import retrofit2.Response
+
+class MainActivity : AppCompatActivity(), Callback<ResponseBody> {
```
Вешать все подряд интерфейсы подряд такое себе решение. Когда их будет дохера (а их будет жохера в большой апке), то ты хер найдешь где находится реализация этого интерфейса по строчкам setListener(this).
Лучше сделать вложенный класс, который реализует этот интерфейс. Он по дефолту будет ссылаться внешний класс, т.е. MainActivity и ты сможешь иметь из него доступ ко всем его полям и методам. И с этим будет проще понять какой нибудь setListener(retrofitListener). Главное не забывать отписывать все эти листенеры (или релизить объекты на которые вешал их), потому что они ссылаются на MainActivity, что может привести к out oof memory

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/PictureAdapter.kt:
```diff
@@ -1,37 +1,21 @@
-    companion object {
-        private const val TAG = "PictureAdapter"
+    fun addItem(image: Bitmap) {
+        items.add(image)
```
будешь долго листать крашнешься на переполнении памяти

rclongloaf replied:
Нужно ограничить количество хранимых картинок

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,70 @@
-        recyclerView.adapter = PictureAdapter()
-        recyclerView.layoutManager = LinearLayoutManager(this)
+    override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
+        pictureAdapter.addItem(BitmapFactory.decodeStream(response.body()?.byteStream()))
```
почему не перенести логику запроса новых картинок в адаптер? а тут писать adapter.loadMore()

rclongloaf replied:
хотя это спорный вопрос. может мы потом захотим где нибудь ещё загрузчик картинок использовать вне адаптера, либо вместо загрузчика брать из других мест. видимо логичнее будет сделать отдельно, чтоб разделить функционал

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,70 @@
+        onLoadMore()
+    }
+
+    fun onLoadMore() {
```
префикс on обычно пишут, если это событие случилось из-за какого то ивента, а тут наоборот ты просишь загрузить больше. больше подходит просто loadMore()

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,70 @@
+        })
+        onLoadMore()
+        onLoadMore()
+        onLoadMore()
```
for (1..MIN_IMAGE_ADAPTER_SIZE)

rclongloaf replied:
и почему 3?

masya38ru replied:
> и почему 3?
цифра добыта эмпирическим путем

rclongloaf replied:
я щас ещё подумал, а у тебя нет бага, что на самом старте у тебя спамится onLoadMore, из-за того, что ты листаешь ещё не прогруженный список? (баг в ифе скролллистенера)

masya38ru replied:
> я щас ещё подумал, а у тебя нет бага, что на самом старте у тебя спамится onLoadMore, из-за того, что ты листаешь ещё не прогруженный список? (баг в ифе скролллистенера)

у меня список стянут в 0 пикселей и пока пикчи не залетят его не прокрутишь)

rclongloaf replied:
но баг всё ещё есть. сделай так, что работал только один loadMore в один момент времени, и чтобы грузил штук по 10 урлов сразу. а потом уже вешай крутилки на итемы, пока не загрузятся сами пикчи

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,70 @@
+    private var firstVisibleItem: Int = 0
+
+    companion object {
+        const val TAG = "MainActivity"
```
лучше в самый верх перенести

rclongloaf replied:
хотя не, везде встречаю companion object в самом низу

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,70 @@
-    lateinit var recyclerView: RecyclerView
+    private var visibleItemCount: Int = 0
+    private var totalItemCount: Int = 0
+    private var firstVisibleItem: Int = 0
```
все 3 переменные можно вынести в метод

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,70 @@
-class MainActivity : AppCompatActivity() {
+    private lateinit var recyclerView: RecyclerView
+    private lateinit var pictureAdapter: PictureAdapter
+    private lateinit var layoutManager: LinearLayoutManager
```
recyclerView и layoutManager можно вынести в метод

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,70 @@
+                    totalItemCount = layoutManager.itemCount
+                    firstVisibleItem = layoutManager.findFirstVisibleItemPosition()
+
+                    if (visibleItemCount + firstVisibleItem >= totalItemCount) {
```
это ведь можно и в адаптере было сделать в onBindViewHolder() { if (position == getItemCount() - 1) ..}

rclongloaf replied:
или в onAttachView, но лучше в onBind, шоб заранее прогружаться начинал

rclongloaf replied:
и ещё тут баг, когда ты можешь успеть быстро свайпать туда сюда в конце списка, то loadMore может сработать несколько раз. таким образом с плохим инетом, можно настакать себе бесконечную очередь загрузки

rclongloaf replied:
про перенос в адаптер передумал, выше уже отписал почему

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,70 @@
+    }
+
+    fun onLoadMore() {
+        RestClient.instance().loadImage().enqueue(this)
```
не вижу нигде отмену загрузки картинок при выходе из приложения

masya38ru committed:
bug fixes:
storage limit,
onLoadMore() -> loadMore()
multiple operation loadMore() fix
cancel loading on closing the app

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,93 @@
+    }
+
+    private lateinit var pictureAdapter: PictureAdapter
+    private val responses = LinkedList<Response<ResponseBody>>()
```
ArrayList плачет в сторонке

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,93 @@
+    fun loadMore(callback: Callback<ResponseBody>) {
+        val call = RestClient.instance().loadImage()
+        for (i in 1..PictureAdapter.MIN_SIZE) {
+            calls.add(call.clone())
```
что тут происходит?

masya38ru replied:
сохраняю вызовы чтоб при закрытии аппки их можно было отменить

rclongloaf replied:
достаточно хранить один Call, т.к. в один момент в идеале должен быть только один процесс загрузки картинок

rclongloaf replied:
и называй их не просто call, а loadImageCall хотя бы

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,93 @@
+        }
+    }
+
+    inner class CallBackImplementer : Callback<ResponseBody> {
```
LoadImageCallback*

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,93 @@
+        super.onStop()
+        for (call in calls)
+            if (!call.isExecuted)
+                call.cancel()
```
на onStart надо бы возобновлять подгрузку списка

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/PictureAdapter.kt:
```diff
@@ -1,37 +1,30 @@
 
     companion object {
-        private const val TAG = "PictureAdapter"
+        const val MAX_SIZE = 500
```
у фреско есть свой кэш картинок, так что их хранение можешь оставить ему

rclongloaf replied:
можешь даже попробовать ручками этот кэш понастраивать

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/PictureAdapter.kt:
```diff
@@ -51,33 +44,15 @@ class PictureAdapter :
+    override fun getItemCount() = itemUrls.size
 
     override fun onBindViewHolder(holder: PictureHolder, position: Int) {
+        val index = position % MAX_SIZE
```
втф, у нас же бесконечный список, а не зацикленный

rclongloaf replied:
аа, ты кэш покругу пустил. ок, избавься от него и забудь

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/PictureAdapter.kt:
```diff
@@ -51,33 +44,15 @@ class PictureAdapter :
-                }
-            })
-        }
+        holder.picture.setImageBitmap(items[index])
```
setImageUrl достаточно будет. проверь только, что кэш фреско работает

masya38ru committed:
changelist:
loading images using fresco
collection types changed
loadImage -> getRandomImageUri
CallBackImplementer -> getImageUriCallback

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,109 @@
         setContentView(R.layout.activity_main)
 
-        recyclerView = findViewById(R.id.recycler_view)
+        Fresco.initialize(this)
```
перенеси лучше в Application класс, чтобы фреско работало нормально со всем приложением, а не только с этим активити

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,109 @@
+        getImageUrlCalls.removeAll(getImageUrlCalls)
+        for (i in 1..PictureAdapter.MIN_SIZE){
+            val getImageUrl = RestClient.instance().getRandomImageUri()
+            getImageUrl.enqueue(callback)
```
сначала закинь в хешсет, а затем вызывай enqueue

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,109 @@
+            getImageUrlCalls.remove(call)
+            val getImageUrl = RestClient.instance().getRandomImageUri()
+            getImageUrl.enqueue(GetImageUriCallback())
+            getImageUrlCalls.add(getImageUrl)
```
копипаста, вынеси эти 3 строчки в отдельный метод

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,109 @@
+        loadMore(GetImageUriCallback())
+    }
+
+    fun loadMore(callback: Callback<ResponseBody>) {
```
private fun, и в других местах тоже private нужен

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,109 @@
-        recyclerView.layoutManager = LinearLayoutManager(this)
+    override fun onDestroy() {
+        super.onDestroy()
+        Fresco.getImagePipeline().clearCaches()
```
это точно нужно?

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,109 @@
+        for (call in getImageUrlCalls) {
+            call.cancel()
+        }
+        getImageUrlCalls = HashSet()
```
всё ещё не вижу onStart

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,109 @@
+        override fun onResponse(call: Call<ResponseBody>, response: Response<ResponseBody>) {
+            getImageUrlCalls.remove(call)
+            responses.add(response)
+            if (responses.size == PictureAdapter.MIN_SIZE) {
```
getImageUrlCalls.isEmpty

rclongloaf commented on app/src/main/java/bonch/dev/myapplication/MainActivity.kt:
```diff
@@ -1,21 +1,109 @@
+
+    private lateinit var pictureAdapter: PictureAdapter
+    private val responses = ArrayList<Response<ResponseBody>>()
+    private var getImageUrlCalls = HashSet<Call<ResponseBody>>()
```
звучит как название метода get***, loadImageUrlCalls больше подходит

masya38ru committed:
28/07/20 changelist:
added App class
Initializing fresco in App class
getImageUrl, getImageUrlCalls -> loadImageUrl, loadImageUrlCalls
added fun addLoadImageUrlCall()
